### PR TITLE
fix(google,log): Avoid log name overriding

### PR DESCRIPTION
**Pull request by AlexisBRENON**
---

Avoid the use of the generic `name` attribute which is overrode by the dict configurator.

I'm not sure if this change break the backward compatibility (`name` attribute is "public" and maybe someone do some things with it...).

### CODE REVIEW

File: contributing-docs/07_local_virtualenv.rst
Comment (shahar1) [inline contributing-docs/07_local_virtualenv.rst]: While it is a good catch, it's better to fix it in a separate PR as it deviates from this one's scope

File: docs/apache-airflow-providers-google/logging/stackdriver.rst
Comment (shahar1) [inline docs/apache-airflow-providers-google/logging/stackdriver.rst:53]: Why do we need the extra `/` all of a sudden? (same goes for line 36)

File: airflow/providers/google/cloud/log/stackdriver_task_handler.py
Comment (shahar1) [inline airflow/providers/google/cloud/log/stackdriver_task_handler.py]: I'm not sure that I fully understand the essence of this change - is the use of `self.name` causing a current bug due to its override, or is it solely intended for differentiation to prevent potential future bugs?

At least for the latter case - it's better to make `self.name` a deprecated property as it breaks backward compatibility (see example [here](https://github.com/shahar1/airflow/blob/d4d4ca38cc3a65f3d61bc6baa6a5eef4de740284/airflow/providers/google/cloud/operators/vertex_ai/custom_job.py#L1360)).

**Review by shahar1 — CHANGES_REQUESTED** [review]

File: docs/apache-airflow-providers-google/logging/stackdriver.rst
Comment (AlexisBRENON) [reply docs/apache-airflow-providers-google/logging/stackdriver.rst:53]: As you can see, the documentation states that "The path section of the URL specifies the name of the log". However, the URL `stackdriver://airflow-tasks` (with only two `/`) is parsed with an empty `path` component, resulting in the name of the log being an empty string (and so, fallbacks on the default value). With, the extra `/`, the URL is parsed as expected, using the `airflow-tasks` as the name of the log:

```py
>>> from urllib.parse import urlsplit
>>> urlsplit("stackdriver://airflow-tasks")
SplitResult(scheme='stackdriver', netloc='airflow-tasks', path='', query='', fragment='')
>>> urlsplit("stackdriver:///airflow-tasks")
SplitResult(scheme='stackdriver', netloc='', path='/airflow-tasks', query='', fragment='')
>>> 
```

File: contributing-docs/07_local_virtualenv.rst
Comment (AlexisBRENON) [reply contributing-docs/07_local_virtualenv.rst]: I will move it to another PR :+1:

File: airflow/providers/google/cloud/log/stackdriver_task_handler.py
Comment (AlexisBRENON) [reply airflow/providers/google/cloud/log/stackdriver_task_handler.py]: Sorry, maybe I forgot to add some context ^^
So, Airflow relies on the [`logging.config.dictConfig`](https://docs.python.org/3/library/logging.config.html) method to [setup the logging stack](https://github.com/apache/airflow/blob/a58441ca1b263cae61a5bb653e6839f0dd29b08e/airflow/logging_config.py#L69).
However, during this setup, it iterates through the handlers and [explicitly sets their name](https://github.com/python/cpython/blob/2a4cbf17af19a01d942f9579342f77c39fbd23c4/Lib/logging/config.py#L578):
```py
                for name in sorted(handlers):
                    try:
                        handler = self.configure_handler(handlers[name])
                        handler.name = name
                        handlers[name] = handler
```

So, before this fix:

1. Setup the remote logging through the environment variables `AIRFLOW__LOGGING__REMOTE_LOGGING="true"` and `AIRFLOW__LOGGING__REMOTE_BASE_LOG_FOLDER="stackdriver://host/path"`
2. Airflow instanciates a `StackdriverTaskHandler` with the name of `"path"`
3. **BUT** the `dictConfig` call overrides the name of the handler with the key of the handlers configuration (i.e. [`task`](https://github.com/apache/airflow/blob/a58441ca1b263cae61a5bb653e6839f0dd29b08e/airflow/config_templates/airflow_local_settings.py#L350)).
4. Hence, the next calls to the `emit` method of the handler will generate logs to the wrong destination (`task` instead of `path`).

Changing the field, from `name` to `gcp_log_name` prevents the overriding from the `dictConfig`. For me, it is a bug fix (and not a change to prevent potential future bugs), but as you noted, it breaks the backward compatibility. Should we still pass through a deprecation period or should we update it ASAP ?

Comment (shahar1) [reply airflow/providers/google/cloud/log/stackdriver_task_handler.py]: Thanks for the context :)

@eladkal Given that the bug fix would be breaking due `name` being a reserved keyword, what do we do in this case regarding deprecation?

File: contributing-docs/07_local_virtualenv.rst
Comment (AlexisBRENON) [reply contributing-docs/07_local_virtualenv.rst]: Already fixed by #38226

File: docs/apache-airflow-providers-google/logging/stackdriver.rst
Comment (shahar1) [reply docs/apache-airflow-providers-google/logging/stackdriver.rst:53]: I understand - I'd be happy for someone else's opinion about that, as it could also be relevant to other handlers.

File: airflow/providers/google/cloud/log/stackdriver_task_handler.py
Comment (eladkal) [reply airflow/providers/google/cloud/log/stackdriver_task_handler.py]: I don't think this qualify as breaking change. This is a bug fix. Sometimes bug fixes change behavior when the previous behavior was clearly wrong and undesired.

Can you think of a senario why someone would want the current behavior over the fix?

Comment (shahar1) [reply airflow/providers/google/cloud/log/stackdriver_task_handler.py]: Nope, but it would be nice adding an indicative exception in case someone passes the `name` param

Comment (AlexisBRENON) [reply airflow/providers/google/cloud/log/stackdriver_task_handler.py]: I don't see any "good" use case where someone would want the current behavior instead of the fix.
@shahar1 I don't change the constructor arguments (you can still get an instance using `StackdriverTaskHandler(name="foo", ...)`) but I change the meaning of the instance attribute `name` (up to now it was supposed to be the name of the GCP log, after this fix it will be the name of the Python handler).
I may add some logs on access (read or write) to the `name` attribute, but I think it may be hard to implement cleanly and will be too verbose (as it is a genuine behavior of the `dictConfig` method).

Comment (shahar1) [reply airflow/providers/google/cloud/log/stackdriver_task_handler.py]: Sounds good, I guess that we could live with it :)
Maybe it's worth deprecating the constructor argument `name` in favor of `gcp_log_name` to avoid confusion, and also people could be aware of the breaking change. I leave it up to you and the committers.

**Comment by shahar1** [comment]
Please notice that static tests fail

**Comment by AlexisBRENON** [comment]
I updated the PR to print a warning when using the `name` constructor argument.
However, my test seems to not pass... https://github.com/apache/airflow/actions/runs/8355631792/job/22871292056?pr=38071
It asks for the default application credentials. But I tried to mock the "transport" stuff to avoid all the required authentication.

**Comment by shahar1** [comment]
> I updated the PR to print a warning when using the `name` constructor argument. However, my test seems to not pass... https://github.com/apache/airflow/actions/runs/8355631792/job/22871292056?pr=38071 It asks for the default application credentials. But I tried to mock the "transport" stuff to avoid all the required authentication.

Try to copy similar logic from existing tests

**Comment by AlexisBRENON** [comment]
Seems good. Failing tests don't seem to be linked to my code.

File: airflow/config_templates/airflow_local_settings.py
Comment (eladkal) [inline airflow/config_templates/airflow_local_settings.py:289]: How is this backward compatible?
What happens if user has latest airflow version with older version of Google provider?

Comment (eladkal) [reply airflow/config_templates/airflow_local_settings.py:289]: Ah sorry forgot discussion in https://github.com/apache/airflow/pull/38071#discussion_r1529196456
I think the newsgragment should be updated to reflect why we are going this. It should reflect more than just what is changed.

Comment (AlexisBRENON) [reply airflow/config_templates/airflow_local_settings.py:289]: We can keep both attributes for the moment (but we will have a warning message at instanciation).

Comment (eladkal) [reply airflow/config_templates/airflow_local_settings.py:289]: If there are no valid use cases for current behavior then there is no reason to keep it. It may cause more confusion.

But others might have different view on this
cc @potiuk @Taragolis @hussein-awala

File: newsfragments/38071.bugfix.rst
Comment (potiuk) [inline newsfragments/38071.bugfix.rst]: @VladaZakharova - maybe someone from your team could verify if that kind of change will be breaking or not.

File: tests/providers/google/cloud/log/test_stackdriver_task_handler.py
Comment (moiseenkov) [inline tests/providers/google/cloud/log/test_stackdriver_task_handler.py:79]: If the new approach uses '///', should we apply it here as well?

Comment (AlexisBRENON) [reply tests/providers/google/cloud/log/test_stackdriver_task_handler.py:79]: The assertion at the bottom checks that only the "path" component of the URI is extracted.
The `:///` is a "special" case where the hostname component is empty.

**Review by moiseenkov — APPROVED** [review]

**Comment by potiuk** [comment]
Errors + static checks need to be fixed.

**Comment by AlexisBRENON** [comment]
> Errors + static checks need to be fixed.

Static checks are about single line newsgragment. But [the comment] from @eladkal suggest to add more info than a single line. Is there a way to by-pass this check or should I put the explanation somewhere else ?

**Comment by potiuk** [comment]
Yes. use significant newsfragment.

**Review by shahar1 — APPROVED** [review]
LGTM

**Review by eladkal — APPROVED** [review]

**Pull request merged**
